### PR TITLE
Add test for simple P/D/I hierarchy duplication

### PR DESCRIPTION
**Pull request by pwalczysko**
This is adding the simple P/D/I hierarchy duplication test. It tests whether all 3 objects were duplicated and it tests whether the linkage between the duplicated P-D-I is as expected (only one link should  exist between the new D and any P, and this one must be the new P, only one link shoiuld exist between the new I and any D, and this one must be the new D).
Please let me know if it makes sense.

### CODE REVIEW

**Comment by pwalczysko** [comment]
I think I am missing one part of the test still. I am testing that there is only one, and just one parent for the D/I. But I am not double checking that this parent is the duplicated P/D. Will push another commit.

**Comment by pwalczysko** [comment]
@ximenesuk : Pushed. I think now it does all the tests, hopefully there is not too much redundancy and clumsiness.

File: components/tools/OmeroPy/test/integration/clitest/test_duplicate.py [hunk 1]
```diff
@@ -39,6 +39,28 @@ def duplicate(self, capfd):
         self.cli.invoke(self.args, strict=True)
         return capfd.readouterr()[0]
 
+    def get_dataset(self, iid):
+        """Retrieve all the parent datasets linked to the image"""
+
+        params = omero.sys.ParametersI()
+        params.addId(iid)
+        query = "select d from Dataset as d"
```
Comment (ximenesuk) [inline components/tools/OmeroPy/test/integration/clitest/test_duplicate.py]: You can use the following Python syntax to neaten this up a bit:

```
query = ("select d from Dataset as d"
         " where exists ("
         "..." )
```

ie if you use brackets you can break strings over several lines. I would also use fewer lines but make them longer, as long as the breaks are in sensible places.

File: components/tools/OmeroPy/test/integration/clitest/test_duplicate.py [hunk 2]
```diff
@@ -128,3 +150,55 @@ def testDuplicateSingleObjectDryRunReport(self, capfd):
...
+        self.link(proj, dset)
+        self.link(dset, img)
+
+        self.args += ['Project:%s' % proj.id.val]
+        self.duplicate(capfd)
+
+        pp = omero.sys.ParametersI()
+        pp.addString("name", namep)
+        query = "select obj from Project obj where obj.name=:name"
+        objsp = self.query.findAllByQuery(query, pp)
```
Comment (ximenesuk) [inline components/tools/OmeroPy/test/integration/clitest/test_duplicate.py:170]: Immediately after this line check if there are two Projects. That way if the duplication has failed the test fails at the first point of failure and you don't need to do the rest needlessly.

File: components/tools/OmeroPy/test/integration/clitest/test_duplicate.py [hunk 3]
```diff
@@ -128,3 +150,55 @@ def testDuplicateSingleObjectDryRunReport(self, capfd):
...
+
+        pd = omero.sys.ParametersI()
+        pd.addString("name", named)
+        query = "select obj from Dataset obj where obj.name=:name"
+        objsd = self.query.findAllByQuery(query, pd)
```
Comment (ximenesuk) [inline components/tools/OmeroPy/test/integration/clitest/test_duplicate.py:184]: Then here check (a) there are two datasets and (b) the new dataset is linked to the new project.

File: components/tools/OmeroPy/test/integration/clitest/test_duplicate.py [hunk 4]
```diff
@@ -128,3 +150,55 @@ def testDuplicateSingleObjectDryRunReport(self, capfd):
...
+
+        # Find the Projects linked to the duplicated dataset
+        did = max(objsd[0].id.val, objsd[1].id.val)
+        proj_linked = self.get_project(did)
+
+        pi = omero.sys.ParametersI()
+        pi.addString("name", namei)
+        query = "select obj from Image obj where obj.name=:name"
+        objsi = self.query.findAllByQuery(query, pi)
```
Comment (ximenesuk) [inline components/tools/OmeroPy/test/integration/clitest/test_duplicate.py:204]: Ditto here with Images and Datasets.

**Comment by ximenesuk** [comment]
Thanks @pwalczysko. The test looks valid, my comments are really just about the ordering of assertions.  While we don't universally do it in our tests there is something to be said for making assertions as soon as you reasonably can. That way the test fails as soon as an error is discovered and you don't need to make needless queries which may throw exceptions due to the earlier errors.

I think it would be good here to bundle in a check that the `--report` is good for a hierarchy. You have the new Project, Dataset and Image ids so you should be able to check if the right strings are in the output.

Finally, my PR is now close to being merged so once you have tweaked this we may have to look at opening a PR directly against develop. But we can discuss that when I'm in Dundee.

**Comment by pwalczysko** [comment]
This should solve the problems remarked on by @ximenesuk

**Comment by pwalczysko** [comment]
@ximenesuk : This should add the simple test for the `skiphead` feature as discussed.

File: components/tools/OmeroPy/test/integration/clitest/test_duplicate.py [hunk 5]
```diff
@@ -39,6 +39,27 @@ def duplicate(self, capfd):
...
+                 "(select l from DatasetImageLink as l where "
+                 "l.child.id=:id and l.parent=d.id)")
+        return self.query.findAllByQuery(query, params)
+
+    def get_project(self, did):
+        """Retrieve all the parent projects linked to the dataaset"""
+
+        params = omero.sys.ParametersI()
+        params.addId(did)
+        query = "select p from Project as p"
```
Comment (ximenesuk) [inline components/tools/OmeroPy/test/integration/clitest/test_duplicate.py]: You can do the same sort of tidying here as you did in the method above.

**Comment by ximenesuk** [comment]
Other than the one bit of hql this all looks good.

**Comment by pwalczysko** [comment]
@ximenesuk : The last commit should fix the tidying of the project query.

**Pull request merged**